PYIC-3429: Add validation to radio buttons

AmritSidhu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Added radio button form validation to all core-front pages which have a radio button forms
 - Added nunjucks filter to push more keys into an object  
 - Added a check function to determine whether journey is undefined and set an errorState
 
Added routes for each individual form POST which has the new form validation check
Used the latest translations provided by content for the error messages

Regarding `page-ipv-bank-account-start.njk form validation` I have added a placeholder message which is from the other error messages, this will need to be updated when we get finalised content

### Why did it change

Users where previously able to submit a form without selecting an option which would default to using the first item in the form. 

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-3429](https://govukverify.atlassian.net/browse/PYIC-3429)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed

[PYIC-3429]: https://govukverify.atlassian.net/browse/PYIC-3429?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-3429: Add functionality to display radio form errors

Added nunjucks filter to push more keys into an object
Added a check function to determine whether journey is undefined and set an errorState

AmritSidhu committed:
PYIC-3429: page-ipv-identity-document-start.njk form validation

AmritSidhu committed:
PYIC-3429: Add pageErrorState as template param

Requires future investigation, seems that the switchs falls through to the pyi-technical-unrecoverable logic as there are no previous breaks
Although this works fine it isnt robust

AmritSidhu committed:
PYIC-3429: page-ipv-identity-postoffice-start.njk form validation

AmritSidhu committed:
PYIC-3429: page-f2f-multiple-doc-check.njk form validation

AmritSidhu committed:
PYIC-3429: page-multiple-doc-check.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-escape.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-cri-escape.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-cri-escape-no-f2f.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-f2f-technical.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-suggest-other-options.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-suggest-other-options-no-f2f.njk form validation

AmritSidhu committed:
PYIC-3429: pyi-post-office.njk form validation

AmritSidhu committed:
PYIC-3429: page-ipv-bank-account-start.njk form validation

AmritSidhu committed:
PYIC-3429: Fix Untrusted URL redirection

Added isPathAllowed check for path validation
Throws an error Path is not allowed if the path is not allowed
Constructs a redirect URL and performs redirection using res.redirect(redirectUrl) if conditions are met
Added and updated test scenarios accordingly

AmritSidhu committed:
PYIC-3429: Fix linting issues

DanCorderIPV commented on src/views/ipv/page-ipv-identity-document-start.njk:
```diff
@@ -1,72 +1,100 @@
-  <p class="govuk-body">{{'pages.pageIpvIdentityDocumentStart.content.paragraph' | translate }}</p>
+
+  {% if errorState %}
+  <p class="govuk-body">{{ 'pages.pageIpvIdentityDocumentStart.content.paragraph' | translate }}</p>	    {{ govukErrorSummary({
```
Don't think this \<p> should be here

DanCorderIPV commented on src/app/ipv/middleware.js:
```diff
@@ -441,6 +448,40 @@ module.exports = {
+        isPathAllowed(req.originalUrl, req.method)
+      ) {
+        const redirectUrl = `${req.baseUrl}${req.path}?errorState=true`;
+        return res.redirect(redirectUrl);
```
It doesn't seem right to redirect the user to another URL - can't we just show the error message on the page we are going to render?

DanCorderIPV commented on src/views/ipv/page-f2f-multiple-doc-check.njk:
```diff
@@ -2,53 +2,76 @@
 {% block content %}
-  <h1 class="govuk-heading-l" id="header" data-page="{{googleTagManagerPageId}}">{{'pages.pageF2FMultipleDockCheck.header' | translate }}</h1>
-  <p class="govuk-body">{{'pages.pageF2FMultipleDockCheck.content.paragraph1' | translate | safe }}</p>
+  {% if errorState %}
```
Can we move this block to `base.njk` and somehow pass it the text entries to display?
Maybe in the validation check create a list of errors (currently maximum length of 1 :) ) that contain the relevant bits of text and then have the shared template display the error box if there are any validation errors.

AmritSidhu replied:
Yes this can be moved into the base.njk, however I think it would still be best to keep the errors in the internationalisation files for consistency and makes it easier to maintain. What I can do is move it into the base and then in the html template set the necessary fields. This would reduce it to half the amount of lines in each template

AmritSidhu committed:
PYIC-3429: Refactor errorSummary to be in the base.njk

errorSummary if statement moved to base
use the template setters to define the error text

AmritSidhu committed:
PYIC-3429: Change redirect to use render

This mitigates potential attacks as we define and sanitize the url

AmritSidhu committed:
PYIC-3429: Move the error summary macro into base

blakeyp commented on src/app/ipv/middleware.test.js:
```diff
@@ -735,4 +735,47 @@ describe("journey middleware", () => {
+      await middleware.formRadioButtonChecked(req, res, next);
+
+      expect(res.render).to.not.have.been.called;
+      expect(next).to.have.been.calledOnce;
```
Shouldn't this be the opposite?
render called and next not called

blakeyp replied:
Ah I think the `called` needs to be called 😅  i.e. `expect(res.render).to.have.been.called()` ?

blakeyp commented on src/views/ipv/page-f2f-multiple-doc-check.njk:
```diff
@@ -5,50 +5,64 @@
+
+    {% if errorState %}
+      {% set errorMessageObject = { 'text': 'pages.pageF2FMultipleDockCheck.content.formErrorMessage.errorRadioMessage' | translate } %}
+      {% set radiosConfig = radiosConfig | setAttribute('errorMessage', errorMessageObject) %}
```
Nice ⚡

AmritSidhu committed:
PYIC-3429: Fix tests to use correct expectation
